agent/api/task.go

```diff
@@ -889,16 +892,34 @@ func (task *Task) updateTaskDesiredStatus() {
 // task's desired status
 // Invariant: container desired status is <= task desired status converted to container status
 // Note: task desired status and container desired status is typically only RUNNING or STOPPED
-func (task *Task) updateContainerDesiredStatus() {
+func (task *Task) updateContainerDesiredStatusUnsafe() {
 	for _, c := range task.Containers {
-		taskDesiredStatus := task.GetDesiredStatus()
+		taskDesiredStatus := task.DesiredStatusUnsafe
```

Comment:
It seems like we can extract this line and line num 909 to the higher level method (`task.UpdateDesiredStatus()`) and pass `taskDesiredStatus` as a parameter to both `updateContainerDesiredStatusUnsafe` and `updateResourceDesiredStatusUnsafe` so that the caller doesn't need to worry about taking any locks:

```golang
 func (task *Task) UpdateDesiredStatus() {
 task.updateTaskDesiredStatus()
 taskDesiredStatus := task.GetDesiredStatus()
 task.updateContainerDesiredStatus(taskDesiredStatus)
 task. updateResourceDesiredStatus(taskDesiredStatus)
}

func (task *Task) updateContainerDesiredStatus(taskDesiredStatus api.TaskStatus) {
 ...
}

func (task *Task) updateResourceDesiredStatus(taskDesiredStatus api.TaskStatus) {
 ...
}

```

Unless there's a need to lock the task down when updating container and resource statuses, this seems like a much simpler approach.

Reply:
The intention was to lock task status while updating container and resource statuses. In the suggestion, some method can still modify the status between `task.updateTaskDesiredStatus()` and `taskDesiredStatus := task.GetDesiredStatus()`.
We can have a lock for those two statements.

Reply:
> In the suggestion, some method can still modify the status between `task.updateTaskDesiredStatus()` and `taskDesiredStatus := task.GetDesiredStatus()`.

I think that's fine and something that should be considered normal. What's the concern if that happens?

Reply:
Just inconsistency (not changing statuses according to the task desired status that we want). I can't think of immediate consequences since the only status anything else might change it to is STOPPED. I'll go ahead and make the suggested change.

Reply:
As per offline discussion, leaving this as it is.